File: test/domain/recovery/recovery_filename_test.dart
// Tests for RecoveryFilename.parse — inverse of FileRecoveryRepository._buildStem.
//
// Filename grammar (millisecond, §5.1.2, OQ-M5-02):
//   YYYY-MM-DDTHH-MM-SS-mmmZ[.txt]
//   YYYY-MM-DDTHH-MM-SS-mmmZ-<N>[.txt]
//
// Rejection rules:
//   - 6-digit fractional seconds (microsecond width) → null (foreign format)
//   - Any name not matching the millisecond grammar → null
//   - Empty string → null
//   - Must never throw; always returns DateTime? (UTC) or null.

import 'package:foglietto/domain/recovery/recovery_filename.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('RecoveryFilename.parse', () {
    test(
      'given_validFilenameWithTxtExtension_when_parse_then_returnsCorrectUtcDateTime',
      () {
        final result = RecoveryFilename.parse('2026-06-14T10-30-00-123Z.txt');
        expect(result, equals(DateTime.utc(2026, 6, 14, 10, 30, 0, 123)));
      },
    );

    test(
      'given_validFilenameWithCollisionSuffix_when_parse_then_returnsSameDateTimeAsBase',
      () {
        final result = RecoveryFilename.parse('2026-06-14T10-30-00-123Z-2.txt');
        expect(result, equals(DateTime.utc(2026, 6, 14, 10, 30, 0, 123)));
      },
    );

    test(
      'given_validFilenameWithoutTxtExtension_when_parse_then_returnsCorrectUtcDateTime',
      () {
        final result = RecoveryFilename.parse('2026-06-14T10-30-00-123Z');
        expect(result, equals(DateTime.utc(2026, 6, 14, 10, 30, 0, 123)));
      },
    );

    test('given_notATimestampFilename_when_parse_then_returnsNull', () {
      final result = RecoveryFilename.parse('not-a-timestamp.txt');
      expect(result, isNull);
    });

    test(
      'given_microsecondFilename6DigitFractional_when_parse_then_returnsNull',
      () {
        final result = RecoveryFilename.parse(
          '2026-06-14T10-30-00-123456Z.txt',
        );
        expect(result, isNull);
      },
    );

    test('given_emptyString_when_parse_then_returnsNull', () {
      final result = RecoveryFilename.parse('');
      expect(result, isNull);
    });

    test('given_validFilename_when_parse_then_returnedDateTimeIsUtc', () {
      final result = RecoveryFilename.parse('2026-06-14T10-30-00-123Z.txt');
      expect(result?.isUtc, isTrue);
    });

    test(
      'given_collisionSuffixFilenameWithoutTxtExtension_when_parse_then_returnsCorrectUtcDateTime',
      () {
        final result = RecoveryFilename.parse('2026-06-14T10-30-00-123Z-5');
        expect(result, equals(DateTime.utc(2026, 6, 14, 10, 30, 0, 123)));
      },
    );
  });
}

File: shared/src/commonTest/kotlin/com/paolosantucci/foglietto/shared/recovery/RecoveryFilenameTest.kt
package com.paolosantucci.foglietto.shared.recovery

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

// Port of test/domain/recovery/recovery_filename_test.dart
// Oracle: Dart RecoveryFilename.parse test suite (9 cases)
//
// Accepted grammar:
//   YYYY-MM-DDTHH-MM-SS-mmmZ[.txt]
//   YYYY-MM-DDTHH-MM-SS-mmmZ-<N>[.txt]
//
// Rejection rules:
//   - 6-digit fractional (microsecond width) -> null
//   - malformed / empty -> null
//   - colons in timestamp -> null (filesystem stems never contain colons)
//   - partial/truncated stems -> null
//   - Never throws on any input.

class RecoveryFilenameTest {

    @Test
    fun parse_validMsStem_returnsCorrectInstant() {
        // "2026-06-20T13-04-09-512Z" -> RecoveryInstant(2026,6,20,13,4,9,512)
        val result = RecoveryFilename.parse("2026-06-20T13-04-09-512Z")
        assertEquals(RecoveryInstant(2026, 6, 20, 13, 4, 9, 512), result)
    }

    @Test
    fun parse_stemWithCollisionSuffix_returnsInstantSuffixDiscarded() {
        // "2026-06-20T13-04-09-512Z-1" -> same instant (suffix discarded)
        val result = RecoveryFilename.parse("2026-06-20T13-04-09-512Z-1")
        assertEquals(RecoveryInstant(2026, 6, 20, 13, 4, 9, 512), result)
    }

    @Test
    fun parse_stemWithTxtExtension_returnsCorrectInstant() {
        // "2026-06-20T13-04-09-512Z.txt" -> correct RecoveryInstant
        val result = RecoveryFilename.parse("2026-06-20T13-04-09-512Z.txt")
        assertEquals(RecoveryInstant(2026, 6, 20, 13, 4, 9, 512), result)
    }

    @Test
    fun parse_stemWithCollisionSuffixAndExtension_returnsCorrectInstant() {
        // "2026-06-20T13-04-09-512Z-2.txt" -> correct RecoveryInstant
        val result = RecoveryFilename.parse("2026-06-20T13-04-09-512Z-2.txt")
        assertEquals(RecoveryInstant(2026, 6, 20, 13, 4, 9, 512), result)
    }

    @Test
    fun parse_sixDigitMicrosecondFractional_returnsNull() {
        // 6-digit (microsecond) precision "2026-06-20T13-04-09-512000Z" -> null, never throws (FR-02, R-05)
        val result = RecoveryFilename.parse("2026-06-20T13-04-09-512000Z")
        assertNull(result)
    }

    @Test
    fun parse_emptyString_returnsNull() {
        // "" -> null, never throws (FR-03)
        val result = RecoveryFilename.parse("")
        assertNull(result)
    }

    @Test
    fun parse_garbageFilename_returnsNull() {
        // "garbage.txt" -> null, never throws (FR-03)
        val result = RecoveryFilename.parse("garbage.txt")
        assertNull(result)
    }

    @Test
    fun parse_colonSeparators_returnsNull() {
        // "2026-06-20T13:04:09-512Z" (colons instead of dashes) -> null (FR-03)
        // filename stems never contain colons
        val result = RecoveryFilename.parse("2026-06-20T13:04:09-512Z")
        assertNull(result)
    }

    @Test
    fun parse_truncatedStem_returnsNull() {
        // "2026-06-20T13-04" (partial/truncated) -> null (FR-03)
        val result = RecoveryFilename.parse("2026-06-20T13-04")
        assertNull(result)
    }
}
